File: Hashing/hash_table_separate_chaining.py
# File Name: hash_table_separate_chaining.py

class HashTable:
    def __init__(self, size):
        self.size = size
        self.table = [[] for _ in range(size)]   # Create empty chains

    # Hash Function (Division Method)
    def hash_function(self, key):
        return key % self.size

    # Insert Operation
    def insert(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        if key not in chain:
            chain.append(key)
            print(f"Key {key} inserted at index {index}.")
        else:
            print(f"Key {key} already exists.")

    # Search Operation
    def search(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        if key in chain:
            print(f"Key {key} found at index {index}.")
            return True
        else:
            print(f"Key {key} not found.")
            return False

    # Delete Operation
    def delete(self, key):
        index = self.hash_function(key)
        chain = self.table[index]

        if key in chain:
            chain.remove(key)
            print(f"Key {key} deleted from index {index}.")
        else:
            print(f"Key {key} not found. Cannot delete.")

    # Display Hash Table
    def display(self):
        print("\nHash Table:")
        for i in range(self.size):
            print(f"Index {i} -> {self.table[i]}")


# ------------------ Main Program ------------------

if __name__ == "__main__":
    size = int(input("Enter size of hash table: "))
    ht = HashTable(size)

    while True:
        print("\n1. Insert")
        print("2. Search")
        print("3. Delete")
        print("4. Display")
        print("5. Exit")

        choice = int(input("Enter your choice: "))

        if choice == 1:
            key = int(input("Enter key to insert: "))
            ht.insert(key)

        elif choice == 2:
            key = int(input("Enter key to search: "))
            ht.search(key)

        elif choice == 3:
            key = int(input("Enter key to delete: "))
            ht.delete(key)

        elif choice == 4:
            ht.display()

        elif choice == 5:
            print("Exiting program.")
            break

        else:
            print("Invalid choice. Please try again.")

File: Hashing/HashTableSeparateChaining.java
// File Name: HashTableSeparateChaining.java

import java.util.LinkedList;
import java.util.Scanner;

class HashTableSeparateChaining {

    private LinkedList<Integer>[] table;
    private int size;

    // Constructor
    public HashTableSeparateChaining(int size) {
        this.size = size;
        table = new LinkedList[size];

        for (int i = 0; i < size; i++) {
            table[i] = new LinkedList<>();
        }
    }

    // Hash Function
    private int hashFunction(int key) {
        return key % size;
    }

    // Insert Operation
    public void insert(int key) {
        int index = hashFunction(key);
        if (!table[index].contains(key)) {
            table[index].add(key);
            System.out.println("Key " + key + " inserted at index " + index);
        } else {
            System.out.println("Key already exists.");
        }
    }

    // Search Operation
    public void search(int key) {
        int index = hashFunction(key);
        if (table[index].contains(key)) {
            System.out.println("Key " + key + " found at index " + index);
        } else {
            System.out.println("Key not found.");
        }
    }

    // Delete Operation
    public void delete(int key) {
        int index = hashFunction(key);
        if (table[index].remove(Integer.valueOf(key))) {
            System.out.println("Key " + key + " deleted from index " + index);
        } else {
            System.out.println("Key not found.");
        }
    }

    // Display Hash Table
    public void display() {
        System.out.println("\nHash Table:");
        for (int i = 0; i < size; i++) {
            System.out.println("Index " + i + " -> " + table[i]);
        }
    }

    // Main Method
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter size of hash table: ");
        int size = sc.nextInt();

        HashTableSeparateChaining ht = new HashTableSeparateChaining(size);

        while (true) {
            System.out.println("\n1.Insert\n2.Search\n3.Delete\n4.Display\n5.Exit");
            System.out.print("Enter your choice: ");
            int choice = sc.nextInt();

            switch (choice) {
                case 1:
                    System.out.print("Enter key to insert: ");
                    ht.insert(sc.nextInt());
                    break;

                case 2:
                    System.out.print("Enter key to search: ");
                    ht.search(sc.nextInt());
                    break;

                case 3:
                    System.out.print("Enter key to delete: ");
                    ht.delete(sc.nextInt());
                    break;

                case 4:
                    ht.display();
                    break;

                case 5:
                    System.exit(0);

                default:
                    System.out.println("Invalid choice");
            }
        }
    }
}
